### ISSUE TITLE
[2020.12] PHP Warnings

### ISSUE THREAD

**Issue by MrPetovan** [friendica/friendica#9251]
Please report here any PHP Warning message obtained running the 2020.09 stable or 2020.12-dev versions.

Closes #8474

**Comment by SpcCw** [comment]
```
PHP Warning:  strpos(): Empty needle in /usr/share/nginx/friends/src/Content/PageInfo.php on line 300" while reading response header from upstream, client: [IP REMOVED], server: friends.deko.cloud, request: "POST /receive/public HTTP/1.1", upstream: "fastcgi://unix:/var/run/php/php7.4-fpm.sock:", host: "friends.deko.cloud"

```

**Comment by annando** [comment]
```
PHP Warning:  iconv(): Charset parameter exceeds the maximum allowed length of 64 characters in /src/Util/ParseUrl.php on line 224
```
Addressed in #9272

**Comment by MrPetovan** [comment]
Ha ha, of course. Without the source of the offending page I'm afraid I can't troubleshoot this one.

**Comment by annando** [comment]
It seems to be this:
```
2020-09-24T03:03:53Z worker [INFO]: add_page_info_to_body: fetch page info for body {"body":"RT @[url=https:\/\/twitter.com\/MeinAugenlicht]MeinAugenlicht[\/url]: Guten Morgen, an der #[url=https:\/\/pirati.ca\/search?tag=MachBar]MachBar[\/url] meldet sich heute das EuGH zu Wort:\n\n[url=https:\/\/podcasts.google.com\/?feed=aHR0cHM6Ly93d3cxLndkci5kZS9tZWRpYXRoZWsvYXVkaW8vY29zbW8vZGFpbHktZ29vZC1uZXdzL2RhaWx5LWdvb2QtbmV3cy0xMTIucG9kY2FzdA&ep=14&episode=L2Nvc21vX2RhaWx5X2dvb2RfbmV3c18yMDIwMDkyM181MjgyMDU0OS5tcDM]podcasts.google.com\/?feed=aHR0cHM6\u2026[\/url]\n\n#[url=https:\/\/pirati.ca\/search?tag=DieGuteNachricht]DieGuteNachricht[\/url]","worker_id":"802874d","worker_cmd":"twitter_sync_run"} - {"file":"PageInfo.php","line":45,"function":"searchAndAppendToBody","uid":"b63a99","process_id":13073}
2020-09-24T03:03:53Z worker [NOTICE]: Curl redirect. {"url":"https:\/\/podcasts.google.com\/?feed=aHR0cHM6Ly93d3cxLndkci5kZS9tZWRpYXRoZWsvYXVkaW8vY29zbW8vZGFpbHktZ29vZC1uZXdzL2RhaWx5LWdvb2QtbmV3cy0xMTIucG9kY2FzdA&ep=14&episode=L2Nvc21vX2RhaWx5X2dvb2RfbmV3c18yMDIwMDkyM181MjgyMDU0OS5tcDM","to":"https:\/\/podcasts.google.com\/feed\/aHR0cHM6Ly93d3cxLndkci5kZS9tZWRpYXRoZWsvYXVkaW8vY29zbW8vZGFpbHktZ29vZC1uZXdzL2RhaWx5LWdvb2QtbmV3cy0xMTIucG9kY2FzdA\/episode\/L2Nvc21vX2RhaWx5X2dvb2RfbmV3c18yMDIwMDkyM181MjgyMDU0OS5tcDM?ep=14"} - {"file":"HTTPRequest.php","line":197,"function":"get","uid":"b63a99","process_id":13073}
2020-09-24T03:03:54Z worker [NOTICE]: Curl redirect. {"url":"https:\/\/podcasts.google.com\/?feed=aHR0cHM6Ly93d3cxLndkci5kZS9tZWRpYXRoZWsvYXVkaW8vY29zbW8vZGFpbHktZ29vZC1uZXdzL2RhaWx5LWdvb2QtbmV3cy0xMTIucG9kY2FzdA&ep=14&episode=L2Nvc21vX2RhaWx5X2dvb2RfbmV3c18yMDIwMDkyM181MjgyMDU0OS5tcDM","to":"https:\/\/podcasts.google.com\/feed\/aHR0cHM6Ly93d3cxLndkci5kZS9tZWRpYXRoZWsvYXVkaW8vY29zbW8vZGFpbHktZ29vZC1uZXdzL2RhaWx5LWdvb2QtbmV3cy0xMTIucG9kY2FzdA\/episode\/L2Nvc21vX2RhaWx5X2dvb2RfbmV3c18yMDIwMDkyM181MjgyMDU0OS5tcDM?ep=14"} - {"file":"HTTPRequest.php","line":197,"function":"get","uid":"b63a99","process_id":13073}
2020-09-24T03:03:54Z worker [INFO]: detected charset UTF-8,%3Csvg%20version%3D%271.1%27%20xmlns%3D%27http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%27%20xmlns%3Axlink%3D%27http%3A%2F%2Fwww.w3.org%2F1999%2Fxlink%27%20x%3D%270px%27%20y%3D%270px%27%20enable-background%3D%27new%200%200%205%202%27%20xml%3Aspace%3D%27preserve%27%20viewBox%3D%270%200%205%202%27%20preserveAspectRatio%3D%27none%20slice%27%3E%3Ccircle%20cx%3D%271%27%20cy%3D%271%27%20r%3D%271%27%20fill%3D%27%23e6e6e6%27%2F%3E%3C%2Fsvg%3E {"worker_id":"802874d","worker_cmd":"twitter_sync_run"} - {"file":"ParseUrl.php","line":223,"function":"getSiteinfo","uid":"b63a99","process_id":13073}
```

**Comment by MrPetovan** [comment]
Thanks, I'll have a look at it.

Edit: Found the issue, there is a `<style>` tag which includes this selector:
```
.xcNBHc{background-size:10px 4px;background-repeat:repeat-x;background-image:url('data:image/svg+xml;charset=UTF-8,%3Csvg%20version%3D%271.1%27%20xmlns%3D%27http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%27%20xmlns%3Axlink%3D%27http%3A%2F%2Fwww.w3.org%2F1999%2Fxlink%27%20x%3D%270px%27%20y%3D%270px%27%20enable-background%3D%27new%200%200%205%202%27%20xml%3Aspace%3D%27preserve%27%20viewBox%3D%270%200%205%202%27%20preserveAspectRatio%3D%27none%20slice%27%3E%3Ccircle%20cx%3D%271%27%20cy%3D%271%27%20r%3D%271%27%20fill%3D%27%23e6e6e6%27%2F%3E%3C%2Fsvg%3E');visibility:hidden}.sZwd7c.B6Vhqe .MyvhI{transition:none}
```
Unfortunately, we capture everything that's following `charset=` without stopping at the comma, yet.

Edit: Futzing with the comma was ill-advised, instead we now escape `<style>` and `<script>` blocks before parsing the charset.

**Comment by ghost** [comment]
latest dev

`PHP Warning:  trim() expects parameter 1 to be string, array given in /friendica/src/Model/Item.php on line 1683`

Invalid line number

**Comment by AlfredSK** [comment]
2020.12-dev PHP 7.4

`in_array() expects parameter 2 to be array, null given in .../include/conversation.php on line 1038`

There are also two notices always appearing together with this warning. See [comment](https://github.com/friendica/friendica/pull/9569#issuecomment-731613088) at the PR that has caused that.

Caused by #9569 

Addressed in #9571

**Comment by ghost** [comment]
2020.12-rc

`PHP Warning: Invalid argument supplied for foreach() in /friendica/src/Worker/APDelivery.php on line 86`

Adressed in #9661

**Comment by AlfredSK** [comment]
2020.12-rc PHP 7.4

```
htmlspecialchars() expects parameter 1 to be string, array given in .../view/smarty3/compiled/81f976e98bf997463e0ed279ce6cf83159b52d7b_2.file.filebrowser.tpl.php on line 44
```

**Comment by AlfredSK** [comment]
2020.12-rc PHP 7.4

Getting quite a lot of this now...

```
PHP Warning:  openssl_verify(): supplied key param cannot be coerced into a public key in .../src/Util/Crypto.php on line 64
```

Corresponding error line in Friendica application log:

```
index [WARNING]: Empty key parameter {"callstack":"HTTPSignature::getSigner, Receiver::processInbox, Inbox::rawContent, call_user_func"} - {"file":"Crypto.php","line":62,"function":"rsaVerify","uid":"ec52a9","process_id":2810802}
```

Fixed by https://github.com/friendica/friendica/pull/9668

**Comment by MrPetovan** [comment]
> 2020.12-rc PHP 7.4
> 
> ```
> htmlspecialchars() expects parameter 1 to be string, array given in .../view/smarty3/compiled/81f976e98bf997463e0ed279ce6cf83159b52d7b_2.file.filebrowser.tpl.php on line 44
> ```

@AlfredSK Can you please paste the content of this file around line 44 please?

**Comment by AlfredSK** [comment]
@MrPetovan  Oof. I'm on mobile only atm. I will try...

**Comment by AlfredSK** [comment]
Line 37-44...

```
<?php
$_from = $_smarty_tpl->smarty->ext->_foreach->init($_smarty_tpl, $_smarty_tpl->tpl_vars['path']->value, 'name', false, 'folder');
$_smarty_tpl->tpl_vars['name']->do_else = true;
if ($_from !== null) foreach ($_from as $_smarty_tpl->tpl_vars['folder']->value => $_smarty_tpl->tpl_vars['name']->value) {
$_smarty_tpl->tpl_vars['name']->do_else = false;
?>
                        <li role="presentation"><a href="#" data-folder="<?php echo htmlspecialchars($_smarty_tpl->tpl_vars['folder']->value, ENT_QUOTES, 'UTF-8');?>
"><?php echo htmlspecialchars($_smarty_tpl->tpl_vars['name']->value, ENT_QUOTES, 'UTF-8');?>
```

Sorry. I don't get the line breaks right with my phone, SSH client and tmux.

**Comment by MrPetovan** [comment]
Nevermind, it’s what I have too, I just wanted to make sure ❤️

**Comment by MrPetovan** [comment]
Hum, I can't reproduce this specific error, I'll need the exact URL that triggered that error. I know it isn't possible from the error message alone but otherwise I'm looking for a needle in a haystack.

**Comment by AlfredSK** [comment]
Ha! I sometimes have an error with an empty $needle in $haystack or something... ;-)

But concerning that warning: I didn't saw it in the last days. I will have a look at my older logs later today.

**Comment by MrPetovan** [comment]
It's okay, it isn't urgent.

**Comment by AlfredSK** [comment]
Unfortunately, I don't have the old logs anymore. Logrotate has done its job. :-)
I checked the PHP logs of the last 5 days. Nothing about `htmlspecialchars`.

**Comment by MrPetovan** [comment]
Closed by #9748
